src/Cryptography/Sodium/Bindings/Signing.hs

```diff
@@ -0,0 +1,328 @@
+
+-------------------
+-- Detached Mode --
+-------------------
+
+-- | The 'cryptoSignDetached' function signs the message @m@, whose length is @mlen@ bytes,
+-- using the secret key @secret key@ and puts the signature into @sig@, which can be up to
+-- 'cryptoSignBytes' bytes long.
+-- The actual length of the signature is put into @siglen@ if @siglen@ is not 'Foreign.nullPtr'.
+-- It is safe to ignore @siglen@ and always consider a signature as 'cryptoSignBytes' bytes long;
+-- shorter signatures will be transparently padded with zeros if necessary.
+--
+-- @since 0.0.1.0
+foreign import capi "sodium.h crypto_sign_detached"
+  cryptoSignDetached ::
+    -- | Pointer to the @signature@.
+    Ptr CUChar ->
+    -- | Pointer to the length of the @signature@.
+    Ptr CULLong ->
+    -- | Pointer to the @message@ to sign.
+    Ptr CUChar ->
+    -- | Length of the @message@.
+    CULLong ->
+    -- | Pointer to the @secret key@.
+    Ptr CUChar ->
+    -- | Return code is 0 on success, -1 on error.
+    IO CInt
+
+-------------------------
+-- Multi-part messages --
+-------------------------
+
+-- $mpm
+-- If the message you're trying to sign doesn't fit in memory, then it can be provided as a sequence
+-- of arbitrarily-sized
+-- chunks. This uses the @Ed25519ph@ signature system, which pre-hashes the message. In other words,
+-- what gets signed is not the message itself but its image through a hash function. If the message
+-- can fit in memory and be supplied as a single chunk, then the single-part API should be
+-- preferred.
+--
+-- == Note
+--
+-- @Ed25519ph(m)@ is intentionally not equivalent to @Ed25519(SHA512(m))@. If, for
+-- some reason, you need to pre-hash the message yourself, then use the multi-part
+-- 'Cryptography.Sodium.Bindings.GenericHashing' module and sign the 512-bit output.
+
+
+-- | Opaque tag representing the hash state struct @crypto_sign_state@ used by the C API.
+--
+-- To use a 'CryptoSignState', use 'withSignState'.
+--
+-- @since 0.0.1.0
+data CryptoSignState
+
+-- | Perform an operation with a 'CryptoSignState' of size 'cryptoSignStateBytes' allocated
+-- and deallocated automatically.
+--
+-- ⚠ The return value of 'withSignState' __MUST NOT__ leak the 'CryptoSignState'.
+--
+-- Please refer to the documentation of 'Foreign.allocaBytes' for more operational details.
+--
+-- @since 0.0.1.0
+withSignState :: (Ptr CryptoSignState -> IO a) -> IO a
+withSignState action = do
+  let size = fromIntegral cryptoSignStateBytes
+  allocaBytes size action
+
+-- | The amount of memory needed to store a 'CryptoSignState'.
+--
+-- @since 0.0.1.0
+foreign import capi "sodium.h crypto_sign_statebytes"
+  cryptoSignStateBytes :: CSize
+
+-- | This function initializes the 'CryptoSignState' state.
+--
+-- It must be called before the first 'cryptoSignUpdate' call.
+--
+-- @since 0.0.1.0
+foreign import capi "sodium.h crypto_sign_init"
+  cryptoSignInit ::
+    -- | Pointer to the cryptographic state. Cannot be 'Foreign.nullPtr'.
```

Comment:
Similar 'restatement' issues here, as well as needing a more explicit indication of size, as with SHA512.